src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -3310,70 +3335,284 @@ new_readdir(DIR *dirp)
+		/* env is not found. Need to be appended if present in current process. */
+		if (!env_found[j]) {
+			rc = d_agetenv_str(&env_value, env_list[j]);
+			if (rc == -DER_NONEXIST) {
+				/* Do nthing if env does not exist*/
+				continue;
+			} else if (rc == -DER_NOMEM) {
+				printf("Error: failed to allocate memory in querying env %s.\n",
+				       env_list[j]);
+				goto err_out2;
+			}
+			/* In case d_agetenv_str() returns -DER_SUCCESS, append the env */
+			rc = asprintf(&env_buf_list[j], "%s=%s", env_list[j], env_value);
+			if (rc < 0) {
+				printf("Error: failed to allocate memory for env %s.\n",
+				       env_list[j]);
+				goto err_out2;
+			}
+			new_envp[i] = env_buf_list[j];
+			i++;
+			/* free the buffer allocated by d_agetenv_str() */
+			free(env_value);
+		}
+	}
+
+	/* append NULL pointer to the end of new_envp */
+	new_envp[i] = NULL;
+
+	return new_envp;
+
+err_out2:
+	/* free the memory allocated for all env buffer */
+	for (j = 0; j < ARRAY_SIZE(env_list); j++) {
+		if (env_buf_list[j])
+			free(env_buf_list[j]);
+	}
+	/* free the buffer returned from d_agetenv_str */
+	if (env_value)
+		free(env_value);
+	free(new_preload_str);
+err_out1:
+	free(new_envp);
+err_out0:
+	return (char **)envp;
+}
+
+static void
+reset_daos_env_before_exec(void)
+{
+	/* bash does fork(), then close opened files before exec(),
+	 * so the fd for log file probably is invalid now.
+	 */
+	d_log_disable_logging();
+
+	if (context_reset) {
+		destroy_all_eqs();
+		daos_eq_lib_fini();
+		daos_inited       = false;
+		daos_debug_inited = false;
+		context_reset     = false;
+	}
+}
+
+int
+execve(const char *filename, char *const argv[], char *const envp[])
+{
+	if (next_execve == NULL) {
+		next_execve = dlsym(RTLD_NEXT, "execve");
+		D_ASSERT(next_execve != NULL);
+	}
+	if (!hook_enabled)
+		return next_execve(filename, argv, envp);
+
+	reset_daos_env_before_exec();
+
+	if (!enforce_exec_env)
+		return next_execve(filename, argv, envp);
+
+	return next_execve(filename, argv, pre_envp(envp));
+}
+
+int
+execvpe(const char *filename, char *const argv[], char *const envp[])
+{
+	if (next_execvpe == NULL) {
+		next_execvpe = dlsym(RTLD_NEXT, "execvpe");
+		D_ASSERT(next_execvpe != NULL);
+	}
+	if (!hook_enabled)
+		return next_execvpe(filename, argv, envp);
+
+	reset_daos_env_before_exec();
+
+	if (!enforce_exec_env)
+		return next_execvpe(filename, argv, envp);
+
+	return next_execvpe(filename, argv, pre_envp(envp));
+}
+
+int
+execv(const char *filename, char *const argv[])
+{
+	if (next_execv == NULL) {
+		next_execv = dlsym(RTLD_NEXT, "execv");
+		D_ASSERT(next_execv != NULL);
+	}
+	if (!hook_enabled)
+		return next_execv(filename, argv);
+
+	reset_daos_env_before_exec();
+
+	if (!enforce_exec_env)
+		return next_execv(filename, argv);
+
+	if (next_execve == NULL) {
+		next_execve = dlsym(RTLD_NEXT, "execve");
+		D_ASSERT(next_execve != NULL);
+	}
+	return next_execve(filename, argv, pre_envp(__environ));
```

Comment:
I still think this is unlikely to work, if a env var has been removed then d_agetenv_str() on it would fail.

Reply:
ok. I will add the interception of unsetenv() then. Thank you!

Reply:
It would be simpler and more reliable to save the env vars that might be required at the time that the parent process itself is initialized.

Reply:
Saving the value at initialization might cause inconsistency in application like bash. If "export" is used to update env in bash script, the saved value would be out of dated. Using env does not have such issue.

Reply:
That's true, but if that's the case then it would be set and the saved value wouldn't be needed.

getenv() probably just checks the environ list so having code which checks environ and adds anything that's missing from environ isn't going to achieve much.

Reply:
It would be more reasonable to let the lib/application set default value if an env is not set instead of using a saved value at the initialization phase.

Assuming env_var's default value is 0.
bash runs a bash script with env_var=1.
"unset env_var" is invoked inside bash script to clear env on purpose.
"env_var=1" will still be used in following new processes.

Relying on getenv would be more consistent to me.